### PR TITLE
Add (Viz)Doom environments, take #2

### PR DESCRIPTION
**Pull request by Miffyli**
#115 but with a fresh start. 

Currently has few environments. ViZDoom, being a "visual" learning env, is not the best fit for entity-gym but the agent should still learn _something_.

TODO:
- [x] Add poetry/etc stuff
- [x] Add quick readme docs
- [x] Get positive results in all envs (try out relpos encoding, study params, double-check env)

Single runs with different envs:
- Defend the Center: https://wandb.ai/entity-neural-network/enn-ppo/runs/2wabrli9
- Health gathering: https://wandb.ai/entity-neural-network/enn-ppo/runs/37rfxti0
- Health-gathering supreme: https://wandb.ai/entity-neural-network/enn-ppo/runs/ok8sx1w1 (this one probably fails to learn because walls are not included as part of entities etc.)

### CODE REVIEW

File: enn_zoo/enn_zoo/vizdoom_env/vizdoom_config.ron [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        translation: (
+            reference_entity: "Player",
+            position_features: ["x", "y"],
+            rotation_angle_feature: "angle"
+        ),
+    ),
+    optim: (
+        bs: 1024,
+        lr: 0.00025,
+        update_epochs: 10
```
Comment (cswinter) [inline enn_zoo/enn_zoo/vizdoom_env/vizdoom_config.ron:23]: that's a lot of epochs! 😮

**Review by cswinter — APPROVED** [review]
Thank you, this is very cool! Code looks good to merge once MyPy passes. I think you can fix most of the MyPY errors by adding explicit type annotations (e.g. `self._last_state: Optional[...] =`) and `assert isinstance(x, Type)` statements. If nothing else works, you can also just append ` # type: ignore` to a line.

Two questions:
1. Is there a configuration of a hard version of the environment that would allow for a reasonable fair comparison to current SotA?
2. Other than the MyPy types, did you run into any pain points or gaps in the documentation?

File: enn_zoo/enn_zoo/vizdoom_env/vizdoom_config.ron (hunk 1 above)
Comment (Miffyli) [reply enn_zoo/enn_zoo/vizdoom_env/vizdoom_config.ron:23]: I used SB3 default PPO settings as a baseline, which seem to work well for ViZDoom when using images. Kinda seems to work here, but indeed makes training slower :D

**Comment by Miffyli** [comment]
1) Hmm health gathering supreme (HGS) is a more difficult environment which appears in papers using ViZDoom (e.g. [Fig. 6](https://arxiv.org/pdf/2006.11751.pdf) in the sample-factory paper). I did not perform any hyperparam tuning so you can probably squeeze more performance out of this. I also think it won't be a fair comparison by any means, as this code has access to exact locations of objects, but at the same time does not know where walls are 😅 .

2) I did struggle quite a bit making this env tbh, but that is because I am not in the mind-set of "multiple entities being controlled". My mind was in "single, controlled player and bunch of other entities". For example, all the action masking and "ids" (which are passed in Observation) are opaque to me and I do not know where they are needed (probably not needed for ViZDoom)?  **However**, I just realized there was a `tutorial.md` available under enn-gym, so that would have helped me out probably :D

**Comment by cswinter** [comment]
Thanks for the feedback! It could be worth adding a special case for environments with a single actor since that's a pretty common use case and could remove a fair bit of complexity. You're actually using the `translate` correctly, but we could certainly use more (any!) docs on all the hyperparameters.

It looks like "defend the center" is doing pretty OK already, it's a bit hard to tell but it looks like the paper baselines take around 20M to reach the same performance as your run reached in 0.5M. Is that a good baseline to compare to? I suppose seeing through walls would still give an unfair advantage on that task as well?

Is there any way to get the wall entities, or are those just not exposed by the environment? I suppose with the difference in visibility you can't make a good comparison anyway, but still nice to see that things are basically working.

**Comment by vwxyzjn** [comment]
Hey, @Miffyli I just fought with mypy recently so tried something :)

File: enn_zoo/pyproject.toml [hunk 2]
```diff
@@ -12,6 +12,7 @@ orjson = "^3.6.5"
 types-requests = "^2.27.7"
 hyperstate = "^0.2.8"
 enn_ppo = {path = "../enn_ppo", develop = true}
+vizdoom = "^1.1.11"
```
Comment (vwxyzjn) [inline enn_zoo/pyproject.toml:15]: Could you also add the updated lock file `poetry.lock`?

**Review by vwxyzjn — COMMENTED** [review]
I can also run the example script in the readme! Yay  :D

```
poetry run python enn_zoo/enn_zoo/train.py --config enn_zoo/enn_zoo/vizdoom_env/vizdoom_config.ron
global_step=2048 charts/episodic_return=4.7621621621621095  charts/episodic_length=41.62162162162162  charts/episodes=37  meanrew=-1.3804688453674316
SPS: 239
```
Just a minor comment on the installation instruction, other things LGTM!

File: enn_zoo/enn_zoo/vizdoom_env/vizdoom_environment.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            action = action.name
+            if "DELTA" in action:
+                self._action_space[action] = CategoricalActionSpace(
+                    DELTA_SPEED_STEPS_STRS
+                )
+            else:
+                self._action_space[action] = CategoricalActionSpace(["off", "on"])
+            self._action_mask[action] = CategoricalActionMask(actor_ids=[0], mask=None)
+
+    def obs_space(self) -> ObsSpace:  # type: ignore
```
Comment (cswinter) [inline enn_zoo/enn_zoo/vizdoom_env/vizdoom_environment.py:138]: The reason it doesn't like this is probably that the parent method is a `classmethod`. You can just remove this and the action_space method implementations and replace all references with `self.__class__.action_space/obs_space`.

Comment (vwxyzjn) [reply enn_zoo/enn_zoo/vizdoom_env/vizdoom_environment.py:138]: image

Do you mean like this?

**Comment by Miffyli** [comment]
> It looks like "defend the center" is doing pretty OK already, it's a bit hard to tell but it looks like the paper baselines take around 20M to reach the same performance as your run reached in 0.5M. Is that a good baseline to compare to? I suppose seeing through walls would still give an unfair advantage on that task as well?

Yup I am going to say it is unfair, as the player is able to see behind him in this setup (entities) vs. vision based :D.

> Is there any way to get the wall entities, or are those just not exposed by the environment? I suppose with the difference in visibility you can't make a good comparison anyway, but still nice to see that things are basically working.

Walls are exposed as well, as some sort of list of points and lines. I was planning to add them but figured the agent would have too hard time learning them, there are too many points/lines defining walls (well in hundreds) and in static environments the agent _should_ be able to learn which parts of the map it can bypass (although constantly updated normalization may make this difficult)

I feel like a combination of tiny bit of vision + entities would work the best. For example, the horizontal line from the middle of the depth map, so agent knows where walls are. Should I try to add walls or this vision thing? 


> Yeah it doesn't look like these methods are actually used anywhere, so just deleting is sufficient. The create_vizdoom_env already implements the methods to the class so the self.__class__.x = are not necessary.

Hmm I am bit confused: removing `obs_class` etc things fails because we are creating a DoomEntityEnvironment in `create_vizdoom_env`, and making them `@classmethods` fails as well. 

Any possibility to make obs/action_spaces non-classmethods (instance methods?)? I feel it adds an unnecessary layer of complexity for config-based envs like Griddly and ViZDoom ^^

**Comment by Miffyli** [comment]
I added the info of N closest walls (also updated code only give info on N closest objects), and now code runs at reasonable pace _and_ I am getting results similar to my [original HealthGatheringSupreme` experiments](https://arxiv.org/pdf/2004.00980.pdf) (Figure 4, mid row, second from left, blue line) using vision! Again, this is only from one run, but previously I did not get above 4.0 score in any of multiple runs, and this time I got this on my first trial :)

If mypy fix is satisfactory (`type: ignore` on classmethod), feel free to merge. I can fix that, too, if needed.

```
global_step=493568 charts/episodic_return=5.499166666666656  charts/episodic_length=162.5  charts/episodes=12  meanrew=0.03999999910593033
SPS: 585
global_step=495616 charts/episodic_return=7.03555555555553  charts/episodic_length=200.88888888888889  charts/episodes=9  meanrew=0.03999999910593033
SPS: 585
global_step=497664 charts/episodic_return=5.799999999999997  charts/episodic_length=170.07692307692307  charts/episodes=13  meanrew=0.03999999910593033
SPS: 585
global_step=499712 charts/episodic_return=6.109999999999979  charts/episodic_length=177.83333333333334  charts/episodes=12  meanrew=0.03999999910593033
SPS: 585
```

image

**Comment by cswinter** [comment]
Very nice, thank you @Miffyli!

**Comment by Miffyli** [comment]
Uh, this never ends :sweat_smile: . ViZDoom has [a list of library dependencies](https://github.com/mwydmuch/ViZDoom#ubuntu). Not sure why the tests passed earlier, though, as it seems that the CI is not configured to have these. 

Doing following would probably work for workflows, but are we ok having vizdoom as a default dependency that will fail unless users install Boost and all that? I feel like it should be optional.

```
- name: Install vizdoom dependencies
  run: sudo apt-get install -y cmake libboost-all-dev libsdl2-dev libfreetype6-dev libgl1-mesa-dev libglu1-mesa-dev libpng-dev libjpeg-dev libbz2-dev libfluidsynth-dev libgme-dev libopenal-dev zlib1g-dev timidity tar nasm
```

**Comment by vwxyzjn** [comment]
In that case, i suggest either making it an optional dependency (e.g, https://github.com/vwxyzjn/cleanrl/blob/master/pyproject.toml#L18-L55) or remove it from poetry and install from pip.
